Tests/Foundation/Tests/TestJSONEncoder.swift

```diff
@@ -869,6 +869,39 @@ class TestJSONEncoder : XCTestCase {
         XCTAssertEqual(JSONEncoder.OutputFormatting.withoutEscapingSlashes.rawValue, 8)
     }
 
+    func test_SR17581_codingEmptyDictionaryWithNonstringKeyDoesRoundtrip() throws {
+        struct Something: Codable {
+            struct Key: Codable, Hashable {
+                var x: String
+            }
+
+            var dict: [Key: String]
+
+            enum CodingKeys: String, CodingKey {
+                case dict
+            }
+
+            init(from decoder: Decoder) throws {
+                let container = try decoder.container(keyedBy: CodingKeys.self)
+                self.dict = try container.decode([Key: String].self, forKey: .dict)
+            }
+
+            func encode(to encoder: Encoder) throws {
+                var container = encoder.container(keyedBy: CodingKeys.self)
+                try container.encode(dict, forKey: .dict)
+            }
+
+            init(dict: [Key: String]) {
+                self.dict = dict
+            }
+        }
+
+        let toEncode = Something(dict: [:])
+        let data = try JSONEncoder().encode(toEncode)
+        let result = try JSONDecoder().decode(Something.self, from: data)
+        XCTAssertEqual(result.dict.count, 0)
+    }
+
     // MARK: - Helper Functions
     private var _jsonEmptyDictionary: Data {
         return "{}".data(using: .utf8)!
```

Comment:
Do we guarantee the empty dictionary is always encoded as `{}` regardless of type? If not, the variable name might be misleading.

Reply:
The empty dictionary is not always encoded as `[]` regardless of type. Specifically, `<T: Encodable> [String: T]` are encoded as JSON objects, and an empty dictionary of that type is encoded as `{}`.